### PR TITLE
Require that someone validate public_name correctly.

### LINKED ISSUE #396: A MUST for inclusion of public_name in outer SNI seems wrong
For at least command line clients (e.g. curl) and perhaps for censorship circumvention tools it makes sense to allow the outer SNI to contain nothing or something that is not the public_name from the SVCB RR. It's possible that any tool that acquires an ECHConfig from a non-DNS source would benefit from that option too.

The offending (for me:-) text is in section 6.1, 2nd list, bullet 7 where it says: 'The value of ECHConfig.contents.public_name MUST be placed in the "server_name" extension.'

There may be clients that do need the above. I'm not sure how to chacterise those in a short phrase, but that's likely possible. At the IETF-110 TLS session mention was made of clients that need to fallback but I'm not sure what reference would be correct for that if e.g. we appended something like "....for clients that need to be able to fallback. [ref]"

### PR DESCRIPTION
**Pull request by chris-wood**
This combines #436 and #447, requiring that validation of public_name happen _somewhere_, providing a reference parser (if one does not exist), and allowing it to happen when parsing ECHConfig.

Closes #396, #405, #424

cc @martinthomson, @davidben, @cjpatton, @sftcd

### CODE REVIEW

**Review by davidben — APPROVED** [review]

**Review by cjpatton — APPROVED** [review]

File: draft-ietf-tls-esni.md [hunk 1]
```diff
@@ -274,25 +280,27 @@ constrain server name lengths. Names may exceed this length if, e.g.,
...
+with an ASCII dot and MUST be parsable as a dot-separated sequence of LDH
+labels, as defined in {{!RFC5890, Section 2.3.1}}. Clients MUST ignore any
+`ECHConfig` structure whose public_name does not meet these criteria. Note
+that these criteria allow IPv4 addresses in dotted-decimal or other notations,
+e.g. {{?RFC3986, Section 7.4}} and {{WHATWG-IPV4}}. Clients SHOULD ignore the
+`ECHConfig` if it contains an encoded IPv4 address. (To determine if a
+public_name value is an IPv4 address, clients can invoke the IPv4 parser
+algorithm in {{WHATWG-IPV4}}. It returns a value when the input is an IPv4
+address.) See {{auth-public-name}} for how the client interprets and validates
+the public_name.
```
Comment (martinthomson) [inline draft-ietf-tls-esni.md:295]: Try splitting this up a little.  You can start multiple sections with ": " to make separate paragraphs (I think).

**Review by martinthomson — APPROVED** [review]

File: draft-ietf-tls-esni.md (hunk 1 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md]: Oh, nice. I didn't know that was possible. I broke up the description into logical units.

File: draft-ietf-tls-esni.md [hunk 2]
```diff
@@ -274,25 +280,32 @@ constrain server name lengths. Names may exceed this length if, e.g.,
...
-criteria are incomplete; they incidentally rule out textual representations of
-IPv6 addresses (see {{!RFC3986, Section 3.2.2}}), but do not exclude IPv4
-addresses in standard dotted-decimal or other non-standard notations such as
-octal and hexadecimal (see {{RFC3986, Section 7.4}}). If `public_name` contains
-a literal IPv4 or IPv6 address, the client SHOULD ignore the `ECHConfig` to
-avoid sending a non-compliant "server_name" extension on the ClientHelloOuter
-(see {{!RFC6066, Section 3}}).
+: The DNS name of the client-facing server, i.e., the entity trusted
+to update the ECH configuration. This is used to correct misconfigured clients,
+as described in {{handle-server-response}}.
```
Comment (ekr) [inline draft-ietf-tls-esni.md:285]: Let's separate out the conformance requirements and the check. I think you want to say "This value MUST be a valid domain name in accordance with XXX".  I believe that would still allow trailing (but not leading) dot, so you may also have to say "The value MUST NOT end with an ASCII dot".

File: draft-ietf-tls-esni.md [hunk 3]
```diff
@@ -274,25 +280,32 @@ constrain server name lengths. Names may exceed this length if, e.g.,
...
+
+: This value MUST NOT begin or end with an ASCII dot and MUST be parsable as a
+dot-separated sequence of LDH labels, as defined in
+{{!RFC5890, Section 2.3.1}}. Clients MUST ignore any `ECHConfig` structure
+whose public_name does not meet these criteria.
+
```
Comment (ekr) [inline draft-ietf-tls-esni.md:291]:
```suggestion
: Clients MUST ignore any `ECHConfig` structure
whose public_name does is not parsable as a
dot-separated sequence of LDH labels, as defined in
{{!RFC5890, Section 2.3.1}} or which
begins or end with an ASCII dot.
```

File: draft-ietf-tls-esni.md [hunk 4]
```diff
@@ -274,25 +280,32 @@ constrain server name lengths. Names may exceed this length if, e.g.,
...
+: These criteria allow IPv4 addresses in dotted-decimal or other notations,
+e.g. {{?RFC3986, Section 7.4}} and {{WHATWG-IPV4}}. Clients SHOULD ignore the
+`ECHConfig` if it contains an encoded IPv4 address. To determine if a
+public_name value is an IPv4 address, clients can invoke the IPv4 parser
+algorithm in {{WHATWG-IPV4}}. It returns a value when the input is an IPv4
+address.
```
Comment (ekr) [inline draft-ietf-tls-esni.md:297]:
```suggestion
:  Clients SHOULD ignore the
`ECHConfig` if it contains an encoded IPv4 address. To determine if a
public_name value is an IPv4 address, clients can invoke the IPv4 parser
algorithm in {{WHATWG-IPV4}}. It returns a value when the input is an IPv4
address.
```

File: draft-ietf-tls-esni.md (hunk 3 above)
Comment (ekr) [reply draft-ietf-tls-esni.md:291]: Note that I think this still allows for bogus names, like ones that are super-long. Do we want to allow clients to check for that?

**Review by ekr — CHANGES_REQUESTED** [review]

File: draft-ietf-tls-esni.md (hunk 3 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:291]: The public_name field is already limited to 255 characters, so I think we're fine.

**Comment by sftcd** [comment]
On 14/06/2021 20:26, Christopher Wood wrote:
> @chris-wood commented on this pull request.
> 
> 
> 
>> +: This value MUST NOT begin or end with an ASCII dot and MUST be parsable as a
> +dot-separated sequence of LDH labels, as defined in
> +{{!RFC5890, Section 2.3.1}}. Clients MUST ignore any `ECHConfig` structure
> +whose public_name does not meet these criteria.
> +
> 
> The public_name field is already limited to 255 characters, so I think we're fine.

The max for a label in a DNS name is 63 octets though isn't
it?

>

**Comment by chris-wood** [comment]
> The max for a label in a DNS name is 63 octets though isn't it?

Yeah, but RFC6066 has no similar requirements (that I can tell), so I don't see we need to impose anything new here. Maybe I'm missing something? And if so, can you please propose text?

File: draft-ietf-tls-esni.md (hunk 2 above)
Comment (chris-wood) [reply draft-ietf-tls-esni.md:285]: Good call -- I believe your suggestions rephrased things to make the requirements and check read easier.

**Comment by sftcd** [comment]
On 14/06/2021 20:33, Christopher Wood wrote:
>> The max for a label in a DNS name is 63 octets though isn't it?
> 
> Yeah, but RFC6066 has no similar requirements (that I can tell), so I
> don't see we need to impose anything new here. Maybe I'm missing 
> something? And if so, can you please propose text?

Fair ask:-)

I guess the public_name mostly has to go in a cert SAN so
would it help to use or refer to rfc5280 text?

The most relevant bit there seems to be:

    When the subjectAltName extension contains a domain name
    system label, the domain name MUST be stored in the
    dNSName (an IA5String). The name MUST be in the "preferred
    name syntax", as specified by Section 3.5 of [RFC1034] and
    as modified by Section 2.1 of [RFC1123].  Note that while
    uppercase and lowercase letters are allowed in domain
    names, no significance is attached to the case.  In
    addition, while the string " " is a legal domain name,
    subjectAltName extensions with a dNSName of " " MUST NOT
    be used.

(That text is near [1]).

That said, that " " is a legal domain name surprises me (and
I'm a co-author of 5280!)

BTW, I'm fine that that all gets fixed later if that's
better.

S.

[1] https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6

**Comment by chris-wood** [comment]
> BTW, I'm fine that that all gets fixed later if that's better.

Sounds good -- I dropped an OPEN ISSUE to note this for later. @sftcd, would you have cycles to take that issue and propose a PR to address it?

**Comment by sftcd** [comment]
On 14/06/2021 20:52, Christopher Wood wrote:
>> BTW, I'm fine that that all gets fixed later if that's better.
> 
> Sounds good -- I dropped an OPEN ISSUE to note this for later.
> @sftcd, would you have cycles to take that issue and propose a PR to
> address it?

Sneaky eh:-)

Sure, I can look at it this week but don't wait on me.

S

>

**Comment by chris-wood** [comment]
> Sure, I can look at it this week but don't wait on me.

Yeah, there's no rush here, I think. This seems like a smaller issue to resolve at some point.

**Pull request merged**